compiler/rustc_lexer/src/lib.rs

```diff
@@ -277,6 +277,62 @@ pub fn strip_shebang(input: &str) -> Option<usize> {
     None
 }
 
+/// Frontmatter is a special attribute type reserved for use by external tools
+///
+/// This must be called after [`strip_shebang`]
+pub fn strip_frontmatter(input: &str) -> Option<usize> {
+    // Whitespace may precede a frontmatter but must end with a newline
+    let mut rest = input;
+    while !rest.is_empty() {
+        let without_spaces = rest.trim_start_matches(|c| is_whitespace(c) && c != '\n');
+        let without_nl = without_spaces.trim_start_matches('\n');
+        if without_nl.len() == rest.len() {
+            // nothing trimmed
+            break;
+        } else if without_nl.len() == without_spaces.len() {
+            // either not a frontmatter or invalid opening
+            return None;
+        }
+        rest = without_nl;
+    }
+
+    // Opens with a line that starts with 3+ `-` followed by an optional identifier
+    const FENCE_CHAR: char = '-';
+    let fence_end =
+        rest.char_indices().find_map(|(i, c)| (c != FENCE_CHAR).then_some(i)).unwrap_or(rest.len());
+    if fence_end < 3 {
+        // either not a frontmatter or invalid frontmatter opening
+        return None;
+    }
+    let (fence_pattern, rest) = rest.split_at(fence_end);
+    let (info, rest) = rest.split_once("\n").unwrap_or((rest, ""));
```

Comment:
Same goes for here, If there are no newlines in the frontmatter string, we'd be sure that this is invalid, so return `None`. Don't try to do additional calculations that are wrong.

Reply:
(also assuming this is related to the discussion at https://github.com/rust-lang/rust/pull/137193#discussion_r1986034704)

Reply:
> nothing will be left to do anything with

which means we should just return `None` here if there is no newline, because we will eventually return `None` anyways. Doing an `unwrap_or` here can make the reader think that we can do some additional work even if there are no newlines, when that is not true.

Reply:
If I'm reading this correct, this `split_once` was removed in https://github.com/rust-lang/rust/compare/1e418eeb4c3b0fdf35889295a8036daa4c084ca6..078a6f68e170033250888604caca3b5ea5a3d4b7 in response to https://github.com/rust-lang/rust/pull/137193#discussion_r1986035639

There is still a line like this
```rust
let (after_closing_fence, rest) = rest.split_once("\n").unwrap_or((rest, ""));
```

In this case, the `unwrap_or` does not represent an error case but unifies the handling of end-of-line with end-of-file.  So it appears there is nothing left to discuss on this point.